lib/srv/switch.go

```diff
@@ -0,0 +1,193 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package srv
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+)
+
+func safeSend(ch chan bool, value bool) (closed bool) {
+	defer func() {
```

Comment:
It should be an error to write to a closed channel. Under what circumstances do we want to try and recover from this? Can we just avoid triggering the error instead?

Reply:
This was a bit of tricky close since only the receiver knew when it wanted to stop. I rewrote to use a callback channel mechanism to avoid relying on `safeSend` to detect when the client wants to stop.